contracts/Paper.sol

```diff
@@ -0,0 +1,157 @@
+            "MUST_OWN_TOKEN_ID"
+        );
+
+        // Further Checks, Effects, and Interactions are contained within the
+        // _claim() function
+        _claim(tokenId, _msgSender());
+    }
+
+    /// @notice Claim Paper for all tokens owned by the sender
+    /// @notice This function will run out of gas if you have too much loot! If
+    /// this is a concern, you should use claimRangeForOwner and claim Dope in
+    /// batches.
+    function claimAllForOwner() external {
+        uint256 tokenBalanceOwner = dopeWarsLootContract.balanceOf(_msgSender());
+
+        // Checks
+        require(tokenBalanceOwner > 0, "NO_TOKENS_OWNED");
+
+        // i < tokenBalanceOwner because tokenBalanceOwner is 1-indexed
+        for (uint256 i = 0; i < tokenBalanceOwner; i++) {
+            // Further Checks, Effects, and Interactions are contained within
+            // the _claim() function
+            _claim(
+                dopeWarsLootContract.tokenOfOwnerByIndex(_msgSender(), i),
+                _msgSender()
+            );
+        }
+    }
+
+    /// @notice Claim Paper for all tokens owned by the sender within a
+    /// given range
+    /// @notice This function is useful if you own too much DWL to claim all at
+    /// once or if you want to leave some Paper unclaimed. If you leave Paper
+    /// unclaimed, however, you cannot claim it once the next season starts.
+    function claimRangeForOwner(uint256 ownerIndexStart, uint256 ownerIndexEnd)
+        external
+    {
+        uint256 tokenBalanceOwner = dopeWarsLootContract.balanceOf(_msgSender());
+
+        // Checks
+        require(tokenBalanceOwner > 0, "NO_TOKENS_OWNED");
+
+        // We use < for ownerIndexEnd and tokenBalanceOwner because
+        // tokenOfOwnerByIndex is 0-indexed while the token balance is 1-indexed
+        require(
+            ownerIndexStart >= 0 && ownerIndexEnd < tokenBalanceOwner,
+            "INDEX_OUT_OF_RANGE"
+        );
+
+        // i <= ownerIndexEnd because ownerIndexEnd is 0-indexed
+        for (uint256 i = ownerIndexStart; i <= ownerIndexEnd; i++) {
+            // Further Checks, Effects, and Interactions are contained within
+            // the _claim() function
+            _claim(
+                dopeWarsLootContract.tokenOfOwnerByIndex(_msgSender(), i),
+                _msgSender()
+            );
+        }
+    }
+
+    /// @dev Internal function to mint Paper upon claiming
+    function _claim(uint256 tokenId, address tokenOwner) internal {
+        // Checks
+        // Check that the token ID is in range
+        // We use >= and <= to here because all of the token IDs are 0-indexed
+        require(
+            tokenId >= tokenIdStart && tokenId <= tokenIdEnd,
+            "TOKEN_ID_OUT_OF_RANGE"
+        );
+
+        // Check that Paper have not already been claimed this season
+        // for a given tokenId
+        require(
+            !claimedByTokenId[tokenId],
```

Comment:
adjusted this and subsequent index into this mapping to not use the extra "season" redirection.